### ISSUE TITLE
`ref_patterns` shouldn't be triggered when sub slices are involved

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust-clippy#16145]
### Description

```rust
#![deny(clippy::ref_patterns)]

fn foo(bar: &[u8]) -> u8 {
    if let &[a, b, ref _rest @ .., c, d, e] = bar {
        a + b + c + d + e
    } else {
        0
    }
}
```

### Version

```text

```

**Comment by profetia** [comment]
Using `&bar` will result in valid code and you do not neeed `*` everywhere:
```rust
fn foo(bar: &[u8]) -> u8 {
    if let [a, b, _rest @ .., c, d, e] = &bar {
        a + b + c + d + e
    } else {
        0
    }
}
```

**Comment by c410-f3r** [comment]
<img width="1411" height="350" alt="Image" src="https://github.com/user-attachments/assets/3b4b54ee-55a3-4a3b-b047-e2cec540fac6" />

Compiles because the standard library implements `Add` for references. De-referencing is still necessary to get owned values.

**Comment by ronnodas** [comment]
Can this be reopened, since the PR was never merged and the lint still triggers on this code? Or should I make a new issue, with roughly the same code block? (Except I am using the individual element bindings in a way that requires dereferencing them.)

**Comment by CommanderStorm** [comment]
It is a bit unclear to me what the exact code (or a minimised version) that causes an false positve here.

This looks like a true positive to me.
The example above looks cleaner when written the way @profetia did.

Could you open an issue with your case?
The reason why is that I think the `restriction` lint (which restrict some part of the language) seems to be working pretty much like I would expect in this case.

This way this can get the attention it needs (and the distinction between the cases is also a bit more clear) and is not burried like this issue.. 😉
I think this could be figured out, but it is much more straightforward with an clear bug report what you are facing.

**Comment by ronnodas** [comment]
@CommanderStorm I guess I'm unsure if this *is* an intended positive, because the lint could have two different intentions:
* Not have any use of the `ref` keyword, in which case this is clearly a true positive.
* Not have uses of the `ref` keyword that can be replaced by match ergonomics, i.e. by inserting or removing some number of `&`/`&mut` on either side of the assignment.

In the second case here is an adapted example that I'd be happy to post as a new issue:
```rust
fn foo(bar: &[u8]) -> u8 {
    if let &[a, b, ref _rest @ .., c, d, e] = bar {
        baz([a, b, c, d, e])
    } else {
        0
    }
}

fn baz(_: [u8; 5]) -> u8 {
    0
}
```
Note that simply removing the `&` and `ref`, as in @profetia's workaround does not work here (I believe this is also what @c410-f3r's last comment is about).

**Comment by CommanderStorm** [comment]
I think this lint is the first, at least from the lint description.
Maybe I am wrong.
We can discuss improvements to lints in an issue too 😉

> What it does
> Checks for usages of the ref keyword.
>
> Why restrict this?
> The ref keyword can be confusing for people unfamiliar with it, and often it is more concise to use & instead.

I think there is space for a third lint beyond the existing `pattern_type_mismatch`, `ref_patterns` (maybe there are more, not sure).
Not sure, needs some thinking about what is wanted why and where 😉 
Likely best as a new issue.

**Comment by c410-f3r** [comment]
`ref_patterns` shouldn't be triggered when sub-slices are involved because of the following use-case: The intention to retrieve **owned** values from pattern destruction when both left and right hand sides have the same type.

AFAICT it is impossible to get an unsized sub-slices from pattern destruction, thus the mandatory need for a "ref"  prefix. In other words, the lint is issuing a warning for a legit scenario, which is a clear error.

```rust
fn foo(bar: &[u8]) -> u8 {
    if let &[a, b, _rest @ .., c, d, e] = bar {
        a + b + c + d + e
    } else {
        0
    }
}
```

> error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:2:20
  |
2 |     if let &[a, b, _rest @ .., c, d, e] = bar {
  |                    ^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
  = note: all local variables must have a statically known size

Feel free to open another issue or create a PR based on the one that was closed.
